middle_end/flambda2/simplify/simplify_common.ml

```diff
@@ -88,9 +87,16 @@ let simplify_projection dacc ~original_term ~deconstructing ~shape ~result_var
     ~result_kind =
   let env = DA.typing_env dacc in
   match T.meet_shape env deconstructing ~shape ~result_var ~result_kind with
-  | Bottom -> Simplified_named.invalid (), TEE.empty, dacc
+  | Bottom ->
+    let dacc = DA.add_variable dacc result_var (T.unknown result_kind) in
```

Comment:
This is one of a few places where I think it makes sense to add a Bottom equation: we know we're in an unreachable branch anyway, so ideally we'd like any `Apply_cont` further down to not be registered for joins. Failing that, at least by marking this variable as Bottom there is a chance that it will slightly improve the result of the join.
Example:
```
type _ tuple_or_function =
| Tuple : (int * int) tuple_or_function
| Function : (int -> int) tuple_or_function

let[@inlined always] apply (type a) (w : a tuple_or_function) (rel : a) x =
  match w with
  | Function -> rel x
  | Tuple -> let _src, dst = rel in dst

let one =
  let x = (fun x -> x + 1) in
  let wit = Sys.opaque_identity Function in
  (apply [@inlined]) wit x 0
```
On this example, on the main branch `one` simplifies to the constant `1`, while on your branch it will still be a dynamically computed value.
And of course, once we stop simplifying invalid branches, there are quite a few more examples (including some less artificial ones) where the types (and as a consequence the code) will be better.

Reply:
Sorry, that was a typo -- it was supposed to say `T.bottom`.